Merge asm/ and analysis/ into arch/ but keep plugin formats.

wargio opened the pull request:
# DO NOT SQUASH
 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This is the first step against rz_arch. this merges still the `librz/asm/` and `librz/analysis/` into the same folder structure but keeps the same formats.

One change that can be noticed is that now we do have a list of all deactivated plugins.

Another big change is also related how we generate the plugins headers, which before needed to be included by the developer each time you add an arch, but now is automatically generated by meson.

After merging this pr, it will be still possible to load external asm and analysis plugins as standalone libraries.

Also I have merged TMS320C64x into TMS320 because it was a duplicated architecture which was available in both tms320 and tms320c64x with code that was duplicated between the two plugins.

Old PR https://github.com/rizinorg/rizin/pull/4120

wargio committed:
Move librz/asm int librz/arch

wargio committed:
Move librz/analysis int librz/arch

wargio committed:
Merge rz_asm and rz_analysis into one library but keep deprecated apis.

The tms320c64x has been merged into tms320.

wargio committed:
Fix license linter

ret2libc commented:
Why removing the common logic in `librz/meson.build`? Is it possible to just declare the plugins in each `librz/*/meson.build` and then construct the header files in `librz/meson.build` so we avoid duplicated meson code?

wargio commented:
> t possible to just declare the plugin

because now it's built as a library, so if we split stuff is easier to handle. also bin plugins has RzBinPlugin and RzBinXtrPlugin so that common logic cannot be easily kept.

XVilka reviewed (changes requested):
Is it possible to not move `librz/analysis/*.c` file? And move only plugins?

XVilka reviewed (commented):
It also broke TMS320C64x analysis:
```
[XX] db/analysis/tms320.c64x_32 arp TMS320
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-434781034/workspace/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'e asm.arch=tms320c64x
e analysis.arch=tms320
arp
' =
-- stdout
--- expected
+++ actual
@@ -26,22 +26,6 @@
 gpr	a13	.32	52 		0
 gpr	a14	.32	56 		0
 gpr	a15	.32	60 		0
-gpr	a16	.32	64 		0
-gpr	a17	.32	68 		0
-gpr	a18	.32	72 		0
-gpr	a19	.32	76 		0
-gpr	a20	.32	80 		0
-gpr	a21	.32	84 		0
-gpr	a22	.32	88 		0
-gpr	a23	.32	92 		0
-gpr	a24	.32	96 		0
-gpr	a25	.32	100 	0
-gpr	a26	.32	104 	0
-gpr	a27	.32	108 	0
-gpr	a28	.32	112 	0
-gpr	a29	.32	116 	0
-gpr	a30	.32	120 	0
-gpr	a31	.32	124 	0
 gpr	b0	.32	128 	0
 gpr	b1	.32	132 	0
 gpr	b2	.32	136 	0
@@ -58,22 +42,6 @@
 gpr	b13	.32	180 	0
 gpr	b14	.32	184 	0
 gpr	b15	.32	188 	0
-gpr	b16	.32	192 	0
-gpr	b17	.32	196 	0
-gpr	b18	.32	200 	0
-gpr	b19	.32	204 	0
-gpr	b20	.32	208 	0
-gpr	b21	.32	212 	0
-gpr	b22	.32	216 	0
-gpr	b23	.32	220 	0
-gpr	b24	.32	224 	0
-gpr	b25	.32	228 	0
-gpr	b26	.32	232 	0
-gpr	b27	.32	236 	0
-gpr	b28	.32	240 	0
-gpr	b29	.32	244 	0
-gpr	b30	.32	248 	0
-gpr	b31	.32	252 	0
 ctr amr     .32 256 0  # Addressing mode register
 ctr csr     .32 260 0  # Control status register
 ctr gfpgfr  .32 264 0  # Galois field multiply control register
@@ -85,22 +53,6 @@
 ctr istp    .32 288 0  # Interrupt service table pointer register
 ctr nrp     .32 292 0  # Nonmaskable interrupt return pointer register
 ctr pce1    .32 296 0  # Program counter, E1 phase
-ctr dier    .32 300 0  # (C64x+ only) Debug interrupt enable register
-ctr dnum    .32 304 0  # (C64x+ only) DSP core number register
-ctr ecr     .32 308 0  # (C64x+ only) Exception clear register
-ctr efr     .32 312 0  # (C64x+ only) Exception flag register
-ctr gplya   .32 316 0  # (C64x+ only) GMPY A-side polynomial register
-ctr gplyb   .32 320 0  # (C64x+ only) GMPY B-side polynomial register
-ctr ierr    .32 324 0  # (C64x+ only) Internal exception report register
-ctr ilc     .32 328 0  # (C64x+ only) Inner loop count register
-ctr itsr    .32 332 0  # (C64x+ only) Interrupt task state register
-ctr ntsr    .32 336 0  # (C64x+ only) NMI/Exception task state register
-ctr rep     .32 340 0  # (C64x+ only) Restricted entry point address register
-ctr rilc    .32 344 0  # (C64x+ only) Reload inner loop count register
-ctr ssr     .32 348 0  # (C64x+ only) Saturation status register
-ctr tsch    .32 352 0  # (C64x+ only) Time-stamp counter (high 32) register
-ctr tscl    .32 356 0  # (C64x+ only) Time-stamp counter (low 32) register
-ctr tsr     .32 360 0  # (C64x+ only) Task state register
 gpr	a0:a1 	.64	364	0
 gpr	a2:a3 	.64	368	0
 gpr	a4:a5 	.64	372	0
@@ -109,14 +61,6 @@
 gpr	a10:a11	.64	384	0
 gpr	a12:a13	.64	388	0
 gpr	a14:a15	.64	392	0
-gpr	a16:a17	.64	396	0
-gpr	a18:a19	.64	400	0
-gpr	a20:a21	.64	404	0
-gpr	a22:a23	.64	408	0
-gpr	a24:a25	.64	412	0
-gpr	a26:a27	.64	416	0
-gpr	a28:a29	.64	420	0
-gpr	a30:a31	.64	424	0
 gpr	b0:b1 	.64	428	0
 gpr	b2:b3 	.64	432	0
 gpr	b4:b5 	.64	436	0
@@ -125,12 +69,4 @@
 gpr	b10:b11	.64	448	0
 gpr	b12:b13	.64	452	0
 gpr	b14:b15	.64	456	0
-gpr	b16:b17	.64	460	0
-gpr	b18:b19	.64	464	0
-gpr	b20:b21	.64	468	0
-gpr	b22:b23	.64	472	0
-gpr	b24:b25	.64	476	0
-gpr	b26:b27	.64	480	0
-gpr	b28:b29	.64	484	0
-gpr	b30:b31	.64	488	0
 



[XX] db/analysis/tms320.c64x_32 320c64 OPEX
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-434781034/workspace/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'e asm.arch=tms320
e asm.cpu=c64x
wx 0041bd10
aoj~{}
' malloc://512280
-- stdout
--- expected
+++ actual
@@ -0,27 +0,15 @@
 [
   {
-    "opcode": "b 0x20de8",
-    "disasm": "b 0x20de8",
-    "pseudo": "goto 0x20de8",
-    "mnemonic": "b",
-    "mask": "ff000000",
-    "jump": 134632,
+    "opcode": "invalid",
+    "disasm": "invalid",
     "sign": false,
     "prefix": 0,
-    "id": 17,
-    "opex": {
-      "operands": [
-        {
-          "type": "imm",
-          "value": 134632
-        }
-      ]
-    },
+    "id": 0,
     "addr": 0,
     "bytes": "0041bd10",
-    "size": 4,
-    "type": "jmp",
+    "size": 0,
+    "type": "ill",
     "esilcost": 0,
     "scale": 0,
     "refptr": 0,
```

XVilka commented:
> Is it possible to not move `librz/analysis/*.c` file? And move only plugins?

Actually, disregard that. I am looking at the tree/code and it looks like keeping everything in `librz/arch/` is better. So the arch can share more logic and data between assembly and analysis, even in the common code, not only particular plugins.

XVilka commented:
Remove both README-analysis.md and README-asm.md - they are empty and useless.

wargio commented:
> It also broke TMS320C64x analysis

yeah. that piece of code was made of `#include "file.c"`

wargio commented:
> It also broke TMS320C64x analysis:

I merged TMS320C64x into TMS320 because it was a dup.

wargio committed:
Removed empty files.

wargio committed:
clang format files that previously were incorrectly excluded.

wargio committed:
Ignore python R1737

This ignores warning 'yield from' directly instead of yielding
each element one by one (use-yield-from)

ret2libc commented on librz/bin/meson.build:
```diff
@@ -141,7 +156,8 @@ rz_bin_sources = [
   'p/bin_xnu_kernelcache.c',
   'p/bin_xtr_fatmach0.c',
-  'p/bin_xtr_pemixed.c',
+  # this was never built.
+  # 'p/bin_xtr_pemixed.c',
```
what do you mean that it was never built? if it was here it should have been built. I don't get it.

wargio replied:
the code was there, but it was never enabled as plugin. that is why when i added it back we had "test failing" but just because now was showing up.

ret2libc reviewed (changes requested):
I don't see why we can't use the approach that was used before to avoid the duplicated meson code. We were already handling RzBinXtr and RzBin plugins. Also, we can probably also just have a single `rz_plugins.h.in` for all the various modules since it is the same.

XVilka reviewed (approved):
LGTM as is. Windows builds and rz-bindgen could be fixed separately after this jumbo PR is merged.

wargio commented:
> I don't see why we can't use the approach that was used before to avoid the duplicated meson code. We were already handling RzBinXtr and RzBin plugins. Also, we can probably also just have a single `rz_plugins.h.in` for all the various modules since it is the same.

You were not handling plugins, but just creating a list of plugins which is easier to do.

ret2libc commented:
> > I don't see why we can't use the approach that was used before to avoid the duplicated meson code. We were already handling RzBinXtr and RzBin plugins. Also, we can probably also just have a single `rz_plugins.h.in` for all the various modules since it is the same.
> 
> You were not handling plugins, but just creating a list of plugins which is easier to do.

What do you mean by "handling"?

In your code I see this:
```meson
rz_lang_plugin_h = configuration_data()
rz_lang_static = []
rz_lang_exports = []
foreach plugin : lang_plugins_list
  rz_lang_static += ['&rz_lang_plugin_' + plugin]
  rz_lang_exports += ['extern RzLangPlugin rz_lang_plugin_' + plugin + ';']
endforeach
rz_lang_plugin_h.set('plugin_static', ', '.join(rz_lang_static))
rz_lang_plugin_h.set('plugin_exports', '\n'.join(rz_lang_exports))
rz_lang_plugin_h_in = configure_file(
  input: 'rz_lang_plugins.h.in',
  output: 'rz_lang_plugins.h',
  configuration: rz_lang_plugin_h
)
```

In the old librz/meson.build there is this:
```meson
  # plugins
  if 'plugins' in module
    foreach plugin : module['plugins']
      l = []
      foreach p : plugin.get('list')
        l += ['&' + plugin.get('base_cname') + p]
      endforeach
      conf_data.set(plugin.get('conf_name'), ', '.join(l))
    endforeach
  endif

...

config_h = configure_file(
  input: 'config.h.in',
  output: 'config.h',
  configuration: conf_data
)

```

What's the difference?

wargio commented:
> What's the difference?

The first one creates ad-hoc file per lib. within the header i have a `#define` and the `export` definition for each plugin.
The second one creates only one header file for all the libs which has only some `#define` in it.

The huge difference (and why this is needed) is that if i follow the old way to generate those, i would need to add the headers of all the libs into the `config.h` just to declare the `export` definitions. in the old code, since it was just a macro, the compiler ignores it unless is used.

So to avoid to have such horrendous `config.h` it is better and cleaner to have one specific header file generated by each library (subfolder) and included only by that library.

ret2libc commented:
You can remove the various changes to the various `librz/**/meson.build` files, they are not necessary.

```
  # plugins
  if 'plugins' in module
    conf_data = configuration_data()
    foreach plugin : module['plugins']
      l = []
      static_plugins = []
      export_plugins = []
      foreach p : plugin.get('list')
        static_plugins += ['&' + plugin.get('base_cname') + p]
        export_plugins += ['extern RzLangPlugin ' + plugin.get('base_cname') + p + ';']
      endforeach
      conf_data.set('plugin_static', ', '.join(static_plugins))
      conf_data.set('plugin_exports', '\n'.join(export_plugins))
    endforeach

    conf_data.set('MODULE_NAME', module_name.to_upper())
    config_h = configure_file(
      input: 'plugins.h.in',
      output: module_name + '_plugins.h',
      configuration: conf_data
    )
  endif
```

`librz/plugins.h.in`:
```c
// SPDX-FileCopyrightText: 2024 RizinOrg <info@rizin.re>
// SPDX-License-Identifier: LGPL-3.0-only

// clang-format off
#ifndef @MODULE_NAME@_PLUGINS_BUILD_H
#define @MODULE_NAME@_PLUGINS_BUILD_H

#define @MODULE_NAME@_STATIC_PLUGINS @plugin_static@

@plugin_exports@
// clang-format on

#endif
```

This generates rz_io_plugins.h, rz_lang_plugins.h, etc. in the `build/librz/` directory (same place where `config.h` was). Then in the various files you just need to do `#include "rz_io_plugins.h"` (instead of `#include "io/rz_io_plugins.h"`). Only difference is the file is not in `librz/io`, but since this header is not installed it doesn't really matter (we had already config.h anyway there).

ret2libc commented:
Tests:
```
cat build-rzarch/librz/rz_lang_plugins.h
// SPDX-FileCopyrightText: 2024 RizinOrg <info@rizin.re>
// SPDX-License-Identifier: LGPL-3.0-only

// clang-format off
#ifndef RZ_LANG_PLUGINS_BUILD_H
#define RZ_LANG_PLUGINS_BUILD_H

#define RZ_LANG_STATIC_PLUGINS &rz_lang_plugin_lib, &rz_lang_plugin_pipe, &rz_lang_plugin_c, &rz_lang_plugin_cpipe

extern RzLangPlugin rz_lang_plugin_lib;
extern RzLangPlugin rz_lang_plugin_pipe;
extern RzLangPlugin rz_lang_plugin_c;
extern RzLangPlugin rz_lang_plugin_cpipe;
// clang-format on

#endif
```
```
cat build-rzarch/librz/rz_io_plugins.h
// SPDX-FileCopyrightText: 2024 RizinOrg <info@rizin.re>
// SPDX-License-Identifier: LGPL-3.0-only

// clang-format off
#ifndef RZ_IO_PLUGINS_BUILD_H
#define RZ_IO_PLUGINS_BUILD_H

#define RZ_IO_STATIC_PLUGINS &rz_io_plugin_ar, &rz_io_plugin_fd, &rz_io_plugin_bfdbg, &rz_io_plugin_bochs, &rz_io_plugin_debug, &rz_io_plugin_default, &rz_io_plugin_dmp, &rz_io_plugin_gdb, &rz_io_plugin_gzip, &rz_io_plugin_http, &rz_io_plugin_ihex, &rz_io_plugin_srec, &rz_io_plugin_mach, &rz_io_plugin_malloc, &rz_io_plugin_null, &rz_io_plugin_procpid, &rz_io_plugin_ptrace, &rz_io_plugin_rzpipe, &rz_io_plugin_rzweb, &rz_io_plugin_rap, &rz_io_plugin_self, &rz_io_plugin_shm, &rz_io_plugin_sparse, &rz_io_plugin_tcp, &rz_io_plugin_winkd, &rz_io_plugin_winedbg, &rz_io_plugin_zip, &rz_io_plugin_qnx

extern RzLangPlugin rz_io_plugin_ar;
extern RzLangPlugin rz_io_plugin_fd;
extern RzLangPlugin rz_io_plugin_bfdbg;
extern RzLangPlugin rz_io_plugin_bochs;
extern RzLangPlugin rz_io_plugin_debug;
extern RzLangPlugin rz_io_plugin_default;
extern RzLangPlugin rz_io_plugin_dmp;
extern RzLangPlugin rz_io_plugin_gdb;
extern RzLangPlugin rz_io_plugin_gzip;
extern RzLangPlugin rz_io_plugin_http;
extern RzLangPlugin rz_io_plugin_ihex;
extern RzLangPlugin rz_io_plugin_srec;
extern RzLangPlugin rz_io_plugin_mach;
extern RzLangPlugin rz_io_plugin_malloc;
extern RzLangPlugin rz_io_plugin_null;
extern RzLangPlugin rz_io_plugin_procpid;
extern RzLangPlugin rz_io_plugin_ptrace;
extern RzLangPlugin rz_io_plugin_rzpipe;
extern RzLangPlugin rz_io_plugin_rzweb;
extern RzLangPlugin rz_io_plugin_rap;
extern RzLangPlugin rz_io_plugin_self;
extern RzLangPlugin rz_io_plugin_shm;
extern RzLangPlugin rz_io_plugin_sparse;
extern RzLangPlugin rz_io_plugin_tcp;
extern RzLangPlugin rz_io_plugin_winkd;
extern RzLangPlugin rz_io_plugin_winedbg;
extern RzLangPlugin rz_io_plugin_zip;
extern RzLangPlugin rz_io_plugin_qnx;
// clang-format on

#endif
```

AFAIS they are the same as the ones you are generating.

ret2libc commented:
> You can remove the various changes to the various `librz/**/meson.build` files, they are not necessary.
> 
> ```
>   # plugins
>   if 'plugins' in module
>     conf_data = configuration_data()
>     foreach plugin : module['plugins']
>       l = []
>       static_plugins = []
>       export_plugins = []
>       foreach p : plugin.get('list')
>         static_plugins += ['&' + plugin.get('base_cname') + p]
>         export_plugins += ['extern RzLangPlugin ' + plugin.get('base_cname') + p + ';']
>       endforeach
>       conf_data.set('plugin_static', ', '.join(static_plugins))
>       conf_data.set('plugin_exports', '\n'.join(export_plugins))
>     endforeach
> 
>     conf_data.set('MODULE_NAME', module_name.to_upper())
>     config_h = configure_file(
>       input: 'plugins.h.in',
>       output: module_name + '_plugins.h',
>       configuration: conf_data
>     )
>   endif
> ```

Commenting to specify that this is a modified code of what was previously in `librz/meson.build`.

ret2libc commented:
> export_plugins += ['extern RzLangPlugin ' + plugin.get('base_cname') + p + ';']

Whoops, this of course needs to be fixed. it should be enough to add a `plugin_type` field to the various `*_plugins` dictionary in `librz/*/meson.build` files.

ret2libc commented on meson.build:
```diff
@@ -795,7 +794,7 @@ summary({
 comb_hash_plugins = []
 in_crca_loop = false
-foreach plugin : hash_plugins.get('list')
+foreach plugin : hash_plugins_list
```
revert

ret2libc commented on meson.build:
```diff
@@ -817,19 +816,17 @@ foreach plugin : hash_plugins.get('list')
-  'IO Plugins': io_plugins.get('list'),
-  'Lang Plugins': lang_plugins.get('list'),
+  'Arch Plugins': arch_plugins_list,
+  'Binary Plugins': bin_plugins_list,
+  'BinXtr Plugins': bin_xtr_plugins_list,
+  'Breakpoint Plugins': bp_plugins_list,
+  'Core Plugins': core_plugins_list,
+  'Crypto Plugins': crypto_plugins_list,
+  'Debug Plugins': debug_plugins_list,
+  'Egg Plugins': egg_plugins_list,
+  'IO Plugins': io_plugins_list,
+  'Lang Plugins': lang_plugins_list,
   'Hash Plugins': comb_hash_plugins,
-  'Parse Plugins': parse_plugins.get('list'),
-  'Demangler Plugins': demangler_plugins.get('list'),
+  'Parse Plugins': parse_plugins_list,
+  'Demangler Plugins': demangler_plugins_list,
 }, section: 'Plugins', list_sep: ', ')
```
revert

ret2libc reviewed (approved):
Better now! Thanks 👍
